napari/layers/_shapes_layer/shape.py

```diff
@@ -0,0 +1,358 @@
+        Color of the shape face
+    opacity : float
+        Opacity of the shape, must be between 0 and 1.
+    z_index : int
+        Specifier of z order priority. Shapes with higher z order are displayed
+        ontop of others.
+    _box : np.ndarray
+        9x2 array of vertices of the bounding box. The first 8 points are
+        the corners and midpoints of the box. The last point is the center
+        of the box
+    _face_vertices : np.ndarray
+        Qx2 array of vertices of all triangles for the shape face
+    _face_triangles : np.ndarray
+        Px3 array of vertex indices that form the triangles for the shape face
+    _edge_vertices : np.ndarray
+        Rx2 array of centers of vertices of triangles for the shape edge.
+        These values should be added to the scaled `_edge_offsets` to get the
+        actual vertex positions. The scaling corresponds to the width of the
+        edge
+    _edge_offsets : np.ndarray
+        Sx2 array of offsets of vertices of triangles for the shape edge. For
+        These values should be scaled and added to the `_edge_vertices` to get
+        the actual vertex positions. The scaling corresponds to the width of
+        the edge
+    _edge_triangles : np.ndarray
+        Tx3 array of vertex indices that form the triangles for the shape edge
+    _shape_types : list
+        List of the supported shape types
+    """
+    _shape_types = ['line', 'rectangle', 'ellipse', 'path', 'polygon']
+
+    def __init__(self, data, shape_type='rectangle', edge_width=1,
+                 edge_color='black', face_color='white', opacity=1, z_index=0):
+
+        self._face_vertices = np.empty((0, 2))
+        self._face_triangles = np.empty((0, 3), dtype=np.uint32)
+        self._edge_vertices = np.empty((0, 2))
+        self._edge_offsets = np.empty((0, 2))
+        self._edge_triangles = np.empty((0, 3), dtype=np.uint32)
+        self._box = np.empty((9, 2))
+
+        self.shape_type = shape_type
+        self.data = np.array(data)
+        self.edge_width = edge_width
+        self.edge_color = edge_color
+        self.face_color = face_color
+        self.opacity = opacity
+        self.z_index = z_index
+
+    @property
+    def shape_type(self):
+        """string: shape_type, must be one of "{'line', 'rectangle', 'ellipse',
+            'path', 'polygon'}".
+        """
+        return self._shape_type
+
+    @shape_type.setter
+    def shape_type(self, shape_type):
+        if shape_type not in self._shape_types:
+            raise ValueError("""shape_type not recognized, must be one of
+                             "{'line', 'rectangle', 'ellipse', 'path',
+                             'polygon'}"
+                             """)
+        self._shape_type = shape_type
+
+    @property
+    def data(self):
+        """np.ndarray: Nx2 array of vertices.
+        """
+        return self._data
+
+    @data.setter
+    def data(self, data):
+        if self.shape_type == 'line':
```

Comment:
ok going to make an overall design comment here. Why not make Shape a base class that each type of shape inherits from. So for example you'd have a separate Rectangle class that has it's own contract upon initialization about what info it needs. Anything that's the same across all shapes should be in the base class. But this individual logic should be in separate subclasses. For example as it is right now there's nothing stopping someone from creating a  rectangle and then calling rectangle.set_type('ellipse').

Reply:
@shanaxel42 I have done this now, please take a look and let me know if you think it is better / if it can be improved in other ways. Thanks!